text/0000-pmgr.md

```diff
@@ -0,0 +1,257 @@
+kernel can also issue a request to the pager to release DRAM pages. The pager is allowed to refuse.
+For this reason, the kernel should not pass ownership of all DRAM to the pager.
+
+### Access to Physical Memory
+
+Since the pager exists in userspace, it does not have access to the DRAM pages that are given to it
+by the kernel. While in most operation, it does not need to read DRAM pages (the pager can submit
+DMA commands without access to the memory), occasionally it might want to read page contents. To do
+this, the pager can simply invent an object ID and submit PageInfo requests to the kernel to
+temporarily map pages into accessable memory.
+
+### Request Types
+
+When communicating lists of pages back and forth, both sides use an array of `PhysRange`.
+The `PhysRange` struct looks like: 
+
+```{rust}
+struct PhysRange {
+    start: u64,
+    len: u32,
+    _flags: u32,
+}
+```
+
+Thus each side can communicate sets of contiguous pages with each other. The flags field is currently unused, but when
+used should be replaced with a newtype that has the same representation as a u32.
+
+## Object Info Requests
+
+The object info struct looks like:
+
+```{rust}
+struct ObjectInfo {
+    lifetime: LifetimeType,
+    backing: BackingType,
+}
+```
+
+This struct will be expanded over-time.
+
+## Eviction
+
+To support eviction, the kernel can inform the pager that a page must by written back to stable
+storage. The pager must honor the kernel's command and must assume the page is dirty. However, this
+is meant to be used as a last resort. The kernel can also submit a *request* that the pager free up
+some number of pages. If a page returned to the kernel is dirty, the kernel will submit an
+eviction notice for that page to inform the pager that it needs to write-back that page.
+
+Note that there is a distinction between "Evict" and "Sync". The former alone requires that data be stored but not
+overwrite object data that may have been modified in-memory. This is required to support transactional semantics well.
+We can combine Sync and Evict to recover standard OS eviction semantics. Sync by itself just syncs and doesn't
+evict. Finally, the pager can request the kernel evict object memory.
+
+## Object Page Requests
+
+The kernel can request the pager to fill a DRAM page with data from stable storage and return it to the kernel. The 
+`ObjectRange` struct looks like: 
+
+```{rust}
+struct ObjectRange {
+    start: u64,
+    len: u32,
+    _flags: u32,
+}
+```
+
+Thus each side can communicate sets of contiguous byte-granularity regions of an object with each other. The flags field is currently unused, but when
+used should be replaced with a newtype that has the same representation as a u32. Note that this means syncing and eviction can
+happen at byte-level granularity, however it's an error to submit a page request that is non-page-aligned. What does
```

Comment:
> however it's an error to submit a page request that is non-page-aligned

couldn't the pager promote this to the closest page boundary? Would that be undesirable behavior?

Reply:
I mean, probably what will happen is that the pager will calculate page number by dividing object-offset by page size, which will then work correctly if non-page-aligned requests are submitted. So yeah, that's probably fine.